rfc/rfc-88/rfc-88.md

```diff
@@ -0,0 +1,601 @@
+          set(row, i, readers[i].read());
+       }
+       return row;
+   }
+   
+   public void setReadHandle(XXReadHandle readHandle) {
+       for (int i = 0; i < readers.length; i += 1) {
+          readers[i].setReadHandle(readHandle);
+       }    
+   }
+   
+   public abstract R createRow(R reuse);
+}
+```
+* Pseudo code for different engines, e.g., Spark and Flink.
+```java
+/**
+ * XFormat struct reader for Spark, which reads data as InternalRow.
+ */
+public class SparkXFormatStructReader extends XFormatStructReader<InternalRow> {
+    private final DataType[] types;
+
+    public SparkXFormatStructReader() {}
+    
+    public InternalRow createRow(InternalRow reuse) {
+        return new GenericInternalRow(types.length);
+    }
+    
+    public void set(InternalRow row, int index, Object object) {
+        row.update(index, object);
+    }
+}
+
+/**
+ * XFormat struct reader for Flink, which reads data as RowData.
+ */
+public class FlinkXFormatStructReader extends XFormatStructReader<RowData> {
+    private final DataType[] types;
+
+    public FlinkXFormatStructReader() {}
+    
+    public RowData createRow(RowData reuse) {
+        return new GenericRowData(types.length);
+    }
+    
+    public void set(RowData row, int index, Object object) {
+        row.setField(index, object);
+    }
+}
+```
+As shown, the fundamental functionality of the format reader can be shared by engines, and they only have to implement 
+their own field setters for rows.
+
+### HoodieRecord & Key Generator Refactoring
+To use the common abstractions of readers and writers, engine-specific rows should be kept in the read/write path. 
+As mentioned above, RFC-46 is a good start to remove the conversion from engine-specific rows to Avro record. Taking 
+these into consideration, the following changes should be made:
+* The engine specific `HoodieRecord` should be implemented for all integrations
+
+    Spark and Hive integration have implemented the new interface proposed by RFC 46, while other engines have not, such 
+    as Flink still uses Avro payload-based `HoodieRecord`. These engines should implement the specific `HoodieRecord` to
+    leverage the new abstraction of reader and writer.
+
+* New abstraction of Key Generator should be implemented
+
+    Currently, the KeyGenerator is tightly coupled with the Avro `GenericRecord`, since Avro will no longer serve as the
+    intermediate representation for Hudi's record payload, it naturally calls for a new abstract of the key generator, 
+    and then each engine integration can customize their own implementations for different scenarios.
+    ```java
+    public interface KeyGenerator<R> {
+        /**
+        * Extract record key from the given engine specific row.
+        */
+        String getRecordKey(R row);
+    
+        /**
+        * Extract partition path from the given engine specific row.
+        */
+        String getPartitionPath(R row)
```

Comment:
Existing interface of KeyGenerator has broken **S**ingle responsibility and **I**nterface segregation principles.

For example, if we want to implement new partitioning logic, we should also implement `String getRecordKey(R row)`, maybe this RFC is a good opportunity to fix it. 

TLDR: In my opinion, computing the record key and partition path is different and should be implemented separately.

Reply:
> Existing interface of KeyGenerator has broken Single responsibility and Interface segregation principles.

I think the reason for mixing up partition and recordKey is that the `KeyGenerator` is actually the generator for `HoodieKey`, which consists of `recordKey` and `partitionPath`. In this RFC, we're not going to change the semantics of `HoodieKey`, nevertheless, I'll make some modification to the `KeyGenerator` interface in this proposal to decrease confusions/misunderstandings, like:

```
public interface KeyGenerator<R> extends Serializable {
 /**
  * Extract HoodieKey from the given engine specific row.
  */
  HoodieKey getKey(R record);
}
```